src/sdc11073/roles/waveformprovider/waveformproviderimpl.py

```diff
@@ -0,0 +1,293 @@
+        """Get next values from generator."""
+        ...
+
+    sample_period: float
+
+
+class _SampleArrayGenerator:
+    """Wraps a waveform generator and makes RtSampleArray objects."""
+
+    def __init__(self, model: AbstractDataModel,
+                 descriptor_handle: str,
+                 generator: WaveformGeneratorProtocol):
+        self._model = model
+        self._descriptor_handle = descriptor_handle
+        self._last_timestamp = None
+        self._activation_state = model.pm_types.ComponentActivation.ON
+        self._generator = generator
+        self.current_rt_sample_array = None
+
+    def set_activation_state(self, component_activation_state: ComponentActivation):
+        """Set activation state of generator.
+
+        If component_activation_state is not "ON", the generator will not generate values.
+        :param component_activation_state: one of pmtypes.ComponentActivation values.
+        """
+        self._activation_state = component_activation_state
+        if component_activation_state == self._model.pm_types.ComponentActivation.ON:
+            self._last_timestamp = time.time()
+
+    def get_next_sample_array(self) -> RtSampleArray:
+        """Read sample values from waveform generator and calculate determination time.
+
+        If activation state is not 'On', no samples are returned.
+        @return: RtSampleArray instance.
+        """
+        if self._activation_state != self._model.pm_types.ComponentActivation.ON:
+            self.current_rt_sample_array = RtSampleArray(
+                self._model, None, self._generator.sample_period, [], self._activation_state)
+        else:
+            now = time.time()
+            observation_time = self._last_timestamp or now
+            samples_count = int((now - observation_time) / self._generator.sample_period)
+            samples = self._generator.next_samples(samples_count)
+            self._last_timestamp = observation_time + self._generator.sample_period * samples_count
+            self.current_rt_sample_array = RtSampleArray(
+                self._model, observation_time, self._generator.sample_period, samples, self._activation_state)
+        return self.current_rt_sample_array
+
+    def set_waveform_generator(self, generator: WaveformGeneratorProtocol):
+        self._generator = generator
+
+    @property
+    def is_active(self) -> bool:
+        return self._activation_state == self._model.pm_types.ComponentActivation.ON
+
+
+class GenericWaveformProvider:
+    """Provide waveform data.
+
+    - runs periodic job to send waveform data
+    """
+
+    DEFAULT_WORKER_THREAD_INTERVAL = 0.1  # seconds
+
+    WARN_LIMIT_REALTIMESAMPLES_BEHIND_SCHEDULE = 0.2  # warn limit when real time samples cannot be sent in time (typically because receiver is too slow)
+    WARN_RATE_REALTIMESAMPLES_BEHIND_SCHEDULE = 5  # max. every x seconds a message
+
+    def __init__(self, mdib: ProviderMdib, log_prefix: str):

```

Comment:
`log_prefix` could be defaulting to `''`

Reply:
done